kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaHistoryManager.java

```diff
@@ -0,0 +1,249 @@
+                // commit file (but there are other later commit files present)
+                throw new RuntimeException(
+                    String.format("No recreatable commits found at %s", logPath));
+            } else {
+                throw new RuntimeException(String.format("No commits found at %s", logPath));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Could not close iterator", e);
+        }
+    }
+
+    /**
+     * Returns an iterator containing a list of files found in the _delta_log directory starting
+     * with {@code startVersion}. Throws a {@link TableNotFoundException} if the directory doesn't
+     * exist or is empty.
+     */
+    private static CloseableIterator<FileStatus> listFrom(
+            TableClient tableClient,
+            Path logPath,
+            long startVersion) throws TableNotFoundException {
+        try {
+            CloseableIterator<FileStatus> files = tableClient
+                .getFileSystemClient()
+                .listFrom(FileNames.listingPrefix(logPath, startVersion));
+            if (!files.hasNext()) {
+                // We treat an empty directory as table not found
+                throw new TableNotFoundException(logPath.getParent().toString()); /* use dataPath */
+            }
+            return files;
+        } catch (FileNotFoundException e) {
+            throw new TableNotFoundException(logPath.toString());
+        } catch (IOException io) {
+            throw new RuntimeException("Failed to list the files in delta log", io);
+        }
+    }
+
+    /**
+     * Returns the commit version and timestamps of all commits starting from version {@code start}.
+     * Guarantees that the commits returned have both monotonically increasing versions and
+     * timestamps.
+     */
+    private static List<Commit> getCommits(TableClient tableClient, Path logPath, long start)
+            throws TableNotFoundException{
+        CloseableIterator<Commit> commits = listFrom(tableClient, logPath, start)
+            .filter(fs -> FileNames.isCommitFile(getName(fs.getPath())))
+            .map(fs -> new Commit(FileNames.deltaVersion(fs.getPath()), fs.getModificationTime()));
+        return monotonizeCommitTimestamps(commits);
+    }
+
+    /**
+     * Makes sure that the commit timestamps are monotonically increasing with respect to commit
+     * versions. Requires the input commits to be sorted by the commit version.
+     */
+    private static List<Commit> monotonizeCommitTimestamps(CloseableIterator<Commit> commits) {
+        List<Commit> monotonizedCommits = new ArrayList<>();
+        long prevTimestamp = Long.MIN_VALUE;
+        long prevVersion = Long.MIN_VALUE;
+        while (commits.hasNext()) {
+            Commit newElem = commits.next();
+            assert(prevVersion < newElem.version); // Verify commits are ordered
+            if (prevTimestamp >= newElem.timestamp) {
```

Comment:
Is this what `delta-spark` also does? If yes, should we have a deisclaimer in the API docs saying that the timestamps should be valid. If the commit files are copied over (or any operation that changes the timestamp of commit files), the API is a best effort to return the snapshot at the given timestamp?

Reply:
Spark (and standalone) does this. I couldn't find a good test in delta-spark for this scenario though. I think if commit files are copied over and timestamps are changed time-travel by version is fully expected to be broken. I'm guessing this is more aimed at commit versions with the same timestamp (since precision is only ms).